src/React.Core/JavaScriptEngineUtils.cs

```diff
@@ -61,6 +77,151 @@ public static bool EnvironmentSupportsClearScript()
+		public static bool TryExecuteFileWithPrecompilation(this IJsEngine engine, ICache cache,
+			IFileSystem fileSystem, string path)
+		{
+			if (!CheckPrecompilationAvailability(engine, cache))
+			{
+				return false;
+			}
+
+			var cacheKey = string.Format(PRECOMPILED_JS_FILE_CACHE_KEY, path);
+			var precompiledScript = cache.Get<IPrecompiledScript>(cacheKey);
+
+			if (precompiledScript == null)
+			{
+				var contents = fileSystem.ReadAsString(path);
+				precompiledScript = engine.Precompile(contents, path);
+				var fullPath = fileSystem.MapPath(path);
+				cache.Set(
+					cacheKey,
+					precompiledScript,
+					slidingExpiration: PRECOMPILED_JS_CACHE_ENTRY_SLIDING_EXPIRATION,
+					cacheDependencyFiles: new[] { fullPath }
+				);
+			}
+
+			engine.Execute(precompiledScript);
+
+			return true;
+		}
+
+		/// <summary>
+		/// Tries to execute a code from embedded JavaScript resource with pre-compilation.
+		/// </summary>
+		/// <param name="engine">Engine to execute a code from embedded JavaScript resource with pre-compilation</param>
+		/// <param name="cache">Cache used for storing the pre-compiled scripts</param>
+		/// <param name="resourceName">The case-sensitive resource name</param>
+		/// <param name="assembly">The assembly, which contains the embedded resource</param>
+		/// <returns>true if can perform a script pre-compilation; otherwise, false.</returns>
+		public static bool TryExecuteResourceWithPrecompilation(this IJsEngine engine, ICache cache,
+			string resourceName, Assembly assembly)
+		{
+			if (!CheckPrecompilationAvailability(engine, cache))
+			{
+				return false;
+			}
+
+			var cacheKey = string.Format(PRECOMPILED_JS_RESOURCE_CACHE_KEY, resourceName);
+			var precompiledScript = cache.Get<IPrecompiledScript>(cacheKey);
+
+			if (precompiledScript == null)
+			{
+				precompiledScript = engine.PrecompileResource(resourceName, assembly);
+				cache.Set(
+					cacheKey,
+					precompiledScript,
+					slidingExpiration: PRECOMPILED_JS_CACHE_ENTRY_SLIDING_EXPIRATION
+				);
+			}
+
+			engine.Execute(precompiledScript);
+
+			return true;
+		}
+
+		/// <summary>
+		/// Checks a availability of the script pre-compilation
+		/// </summary>
+		/// <param name="engine">Instance of the JavaScript engine</param>
+		/// <param name="cache">Cache used for storing the pre-compiled scripts</param>
+		/// <returns>true if the script pre-compilation is available; otherwise, false.</returns>
+		private static bool CheckPrecompilationAvailability(IJsEngine engine, ICache cache)
+		{
+			if (!engine.SupportsScriptPrecompilation)
+			{
+				Trace.WriteLine(string.Format("The {0} version {1} does not support the script pre-compilation.",
+					engine.Name, engine.Version));
+				return false;
+			}
+
+			if (cache is NullCache)
+			{
+				Trace.WriteLine("Usage of script pre-compilation without caching does not make sense.");
+				return false;
+			}
+
+			return true;
+		}
+
```

Comment:
It'd be good to put all this code into a separate file (eg. `JavaScriptPrecompilation.cs`) to keep things organized.